TraderJoe-Bots/Large-Swap/src/multicall2.ts

```diff
@@ -0,0 +1,65 @@
+import { Contract } from "@ethersproject/contracts";
+import { Abi } from "ethers-multicall/dist/abi";
+import { Contract as MulticallContract, ContractCall, Provider, setMulticallAddress } from "ethers-multicall";
+import { Provider as EthersProvider } from "@ethersproject/providers";
+import { MULTICALL2_ABI } from "./utils";
+
+// Provider class heavily based on ethers-multicall, but supporting specifying
+// a blockTag for the multicall contract call.
+
+class MulticallProvider extends Provider {
+  constructor(provider: EthersProvider, multicallAddress: Record<number, string>, chainId?: number) {
+    // clean ethers-multicall Multicall addresses list so a Multicall contract is not called by mistake
+    const ethersMulticallChains = [1, 3, 4, 5, 42, 56, 66, 97, 100, 128, 137, 250, 1337, 42161, 43114, 80001];
+    ethersMulticallChains.forEach((network) => {
+      setMulticallAddress(network, undefined as unknown as string);
+    });
+    for (const [network, address] of Object.entries(multicallAddress)) {
+      // setMulticallAddress(network as unknown as number, address);
+      setMulticallAddress(parseInt(network), address);
+    }
+    super(provider, chainId);
+  }
+
+  public async all<T extends any[] = any[]>(calls: ContractCall[], blockTag?: number | string) {
+    // @ts-ignore
+    if (!this._provider) {
+      throw new Error("Provider should be initialized before use.");
+    }
+
+    // @ts-ignore
+    return this._all<T>(calls, this._multicallAddress, this._provider, blockTag);
+  }
+
+  private async _all<T extends any[] = any[]>(
+    calls: ContractCall[],
+    multicallAddress: string,
+    provider: EthersProvider,
+    blockTag?: number | string
+  ): Promise<T> {
+    const Multicall2 = new Contract(multicallAddress, MULTICALL2_ABI, provider);
+    const requests = calls.map((call) => ({
+      target: call.contract.address,
+      callData: Abi.encode(call.name, call.inputs, call.params),
+    }));
+    const response = await Multicall2.callStatic.tryAggregate(false, requests, { blockTag });
+    const results = [];
+    for (let i = 0; i < calls.length; i++) {
+      const outputs = calls[i].outputs;
+      const result = response[i];
+
+      if (result.success === false) {
+        break;
+      }
+
+      const params = Abi.decode(outputs, result.returnData);
+      results.push({
+        success: result.success,
+        returnData: outputs.length === 1 ? params[0] : params,
+      });
+    }
+    return results as T;
```

Comment:
I think in this case I prefer to use `true` when calling `tryAggregate` at the moment I have no case in mind when I would like to use `false` because from the bots perspective, each time we query the blockchain is because we need some info. This will imply to have a `try/catch` in this section of code.

Another thing is that I would like to return follow the `trySomething` pattern, which return `[bool, ...returnedTypes]` so the first element in the returned data indicates if everything was ok or not. So when doing test calls like fetching tokens as we do in this bot we can have something like:
```typescript
const [success, ...results] = await multicall2.all(listOfCalls, blockNumber);

if(success) {
  // results contains all the data needed
  doChecks(results);
}
else {
  // something fails in the calls
}
```

Reply:
cc: @Hyodar, wdyt?

Reply:
Yes, about using `true` on `tryMulticall`, I was currently working on adding the normal `aggregate` to this `Multicall2`, then separating `all` and `tryAll`. In the former, only the actual call results are returned, so I think it's better in terms of usability when there's no need for the `success` flag.

About the output pattern, we could work on something like that, but `tryAggregate` returns a flag of success for each of the calls, and that could be useful in some cases. So, another option we have that is a bit different from the current approach is returning an array of `success` and an array of `results`.

Reply:
Also, I don't think this function should stop decoding the results if one call is not successful. Whoever is using it should be dealing with `success` later, since one of the main use cases is grouping calls that can fail.

Reply:
Agree with the above comment. Just another note on my suggestion, I know they are returning that `success` flag per call, but the flag I would like to have is to be able to know if something fails, is like having a short circuit flag.

But yes, when success is required it is better to use `aggregate` so having that `all` and `tryAll` would be nice.

I think we need to be agree on what to change in this PR and what left as it is, because I don't want to jeopardize the delivery with technicalities.

Reply:
Or we can keep it as it is this time, because it is working :upside_down_face:, and what we are discussing is how we would like to build this to use it from now on. We can do that later, and maybe add it to GAM.

Reply:
☝️ This is what I was thinking as well. In general, I want to study the `multicall`/`multicall2` approach more, so I have a firmer understanding of how it works, since it is new to me. (I feel I have a good basic understanding, but I'd like to understand it more)

Then that could inform me on how to make it more "generalized" for other use cases. I figured my current implementation works for the bot I currently have, but it would need to be updated and worked on before adding to the `GAM`.